proposed/simple-csharp-programs.md

```diff
@@ -0,0 +1,461 @@
+
+## User experience walkthrough
+
+Given the following code in a singular `Hello.cs` in a directory:
+
+```csharp
+WriteLine("Hello World");
+```
+
+You can just run it:
+
+```bash
+$ dotnet run
+Hello World
+```
+
+Or with a `#!` directive,
+
+```csharp
+#!/path/to/dotnet
+WriteLine("Hello World");
+```
+
+and running:
+
+```bash
+$ ./hello.cs
+Hello World
+```
+
+When running via `dotnet run` or the `#!` directive, artifacts may be created on disk in the temp folder. They could also be generated in memory.
+
+The `using`s that are normally in a C# project template are implicitly available at design-time. This is accomplished with [Global Using Directives](https://github.com/dotnet/csharplang/blob/main/proposals/GlobalUsingDirective.md) that the build tooling will have available.
+
+The `#!` directive is covered in [Hashbang (#!) support](https://github.com/dotnet/csharplang/issues/3507).
+
+### Building instead of running
+
+You can also build it with `dotnet build`:
+
+```bash
+$ dotnet build
+$ ls
+Hello.exe
+$ ./Hello.exe
+Hello World
+```
+
+When built, it creates a single executable that it drops in the current working directory.
+
+### Referencing a package
+
+You can also pull in a package (syntax perhaps subject to change):
+
+```csharp
+#r "nuget: Newtonsoft.Json"
+
+record Person(string Name);
+
+var json = "{\"name\": \"dotnet bot\"}";
+var person = JsonConvert.DeserializeObject<Person>(json);
+
+Console.WriteLine(person.Name);
+```
+
+And it'll run after downloading the package:
+
+```bash
+$ dotnet run
+dotnet bot
+```
+
+Or via a `#!` directive,
+
+```csharp
+#!/path/to/dotnet
+#r "nuget: Newtonsoft.Json"
+
+record Person(string Name);
+
+var json = "{\"name\": \"dotnet bot\"}";
+var person = JsonConvert.DeserializeObject<Person>(json);
+
+Console.WriteLine(person.Name);
+```
+
+and running:
+
+```bash
+$ ./hello.cs
+dotnet bot
+```
+
+The `#r "nuget:..."` syntax is already supported by C# today in .NET Interactive. It is a shared implementation between C# and F#. The syntax and mechanism could be re-used here.
+
+### Coding against a framework/sdk like ASP.NET Core
+
+You can also reference a framework/sdk (nomenclature up for debate) via a `#!` directive:
+
+```csharp
+#!/path/to/dotnet-aspnet
+
+var app = WebApplication.CreateBuilder(args);
+app.MapGet("/", () => "Hello World");
+await app.RunAsync();
+```
+
+The `dotnet-aspnet` gesture indicates two things to the build/run mechanism:
+
+* ASP.NET Core gets loaded and referenced
+* Applicable implicit ASP.NET Core `using`s are brought into scope
+
+And you can run it:
+
+```bash
+$ dotnet run
+Listening on localhost://12345
+```
+
+The syntax for coding against a framework/sdk could also be more similar to the `#r "nuget:"` syntax shown earlier instead of being a `#!` directive.
+
+### Running a specific file
+
+If you have multiple files in the same directory, `dotnet run` won't work by default. You'll need to specify the file you wish to run:
```

Comment:
So if I do `dotnet build` followed by `dotnet run` it won't work anymore because there are now 2 files in the directory?

Reply:
I interpreted that as "multiple _C#_ files in the same directory," presumably identified by the .cs extension. But that would be a good point to clarify.

Reply:
`dotnet run` should, in case it can't identify "entry file", prompt the user to pick/`dotnet run file.ext` with the actual list of files present.

This is in order to still be more helpful than a bland error message.